### PR TITLE
Feature: Implement Haptic Feedback 6.05

### PR DESCRIPTION
**Pull request by Relys**
All function calls should be < 6.05 safe as they explicitly check if the play tone pointers aren't null to determine if the API is available.

Features:
Duty Cycle Solid Tone: Duty Cycle threshold where solid tone is played. Tiltback will pulse at a period of 200ms and then go solid once this value is set. Set <= tiltback to always have it solid. Set >=95 to disable (just like tiltback)
Duty Cycle Strength: Zero disables. Will pulse at 200ms when duty cycle tiltback is present.
Duty Cycle Frequency:
Error Strength: Zero Disables. Will pulse at 200ms with a 400ms pause every 2 beeps for hv/lv. For temp (and bms in future) it will pulse at half rate. i.e. 100ms with 200ms pause.
Error Frequency:
Vibrate Strength: Zero disables. Plays in separate channel
Vibrate Frequency:

### CODE REVIEW

File: src/conf/datatypes.h [hunk 1]
```diff
@@ -185,6 +185,13 @@ typedef struct {
     float tiltback_lv;
     float tiltback_return_speed;
     float tiltback_constant;
+    float haptic_duty_solid;
+    float haptic_duty_volt;
+    int haptic_duty_freq;
+    float haptic_error_volt;
+    int haptic_error_freq;
+    float haptic_vibrate_volt;
+    int haptic_vibrate_freq;
```
Comment (lukash) [inline src/conf/datatypes.h:194]: Can you make a struct member `haptic_feedback` the same way e.g. the LEDs have it's own config structs? Then you can easily make a copy of it in your `HapticFeedback` struct.

File: src/haptic_feedback.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    haptic_feedback->error.peroid = 200;  // 200 ms
+    haptic_feedback->note_duration = 0;
+    haptic_feedback->beep_num = 0;
+}
+
+void haptic_feedback_play_single_tone(float freq, float volt, int channel) {
+    if (!VESC_IF->foc_play_tone) {
+        return;
+    }
+    if ((volt > 0) && (freq > 0.0)) {
```
Comment (lukash) [inline src/haptic_feedback.c:49]: The braces around the `>` expressions are redundant (and this is a very common expression composition). Can you remove them? This repeats multiple times throughout the PR.

When in doubt, refer to [C Operator Precedence](https://en.cppreference.com/w/c/language/operator_precedence.)

File: src/haptic_feedback.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            haptic_feedback->haptic_feedback_state = haptic_feedback->error;
+            switch (state->sat) {
+            case SAT_PB_TEMPERATURE:
+                // Careful
+                if (haptic_feedback->haptic_feedback_state.peroid > 0) {
+                    haptic_feedback->haptic_feedback_state.peroid /= 2;
+                }
+                __attribute__((fallthrough));
+            case SAT_PB_LOW_VOLTAGE:
+                __attribute__((fallthrough));
```
Comment (lukash) [inline src/haptic_feedback.c:94]: The attribute isn't needed if the `case` labels follow directly after each other.

File: src/haptic_feedback.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                break;
+            }
+        }
+        // Reset state if a different type of event is coming in
+        if (haptic_feedback->haptic_feedback_state.type != prev) {
+            haptic_feedback_reset(haptic_feedback, current_time);
+        }
+
+        if (!haptic_feedback->tone_in_progress) {
+            haptic_feedback_play_two_tones(haptic_feedback);
```
Comment (lukash) [inline src/haptic_feedback.c:112]: In a clean design this call shouldn't be here. This Function should just set up the tones or the pattern, and the `haptic_feedback_update()` function take care of the playing by itself. I understand the logic you have relies on this, see my comment on the `haptic_feedback_update()` function.

File: src/haptic_feedback.c [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                    );
+                    haptic_feedback->beep_num--;
+                } else {
+                    haptic_feedback_play_two_tones(haptic_feedback);
+                }
+            } else {
+                VESC_IF->foc_stop_audio(false);
+            }
+        }
+    }
```
Comment (lukash) [inline src/haptic_feedback.c:152]: This logic is very complex and opaque. You also repeatedly call the play functions all the time, which is not very clean. What I'd do instead is take a similar approach to when you implement an animation. I would store a `bool` that says whether you are currently playing a tone. Then, depending on the time, determine whether you're supposed to be playing a tone, and if that differs from the current state, start or stop playing the tone.

For a duty tone pattern, tone-gap-tone-gap-..., you can do e.g.:
```C
const float tone_length = 0.2f;
const float period = tone_length  * 2; // note: according to convention I call period the whole segment that repeats
time = fmodf(time, period); // now time is your time within a single period
bool should_be_playing = time < tone_length;
```

For lets say three beeps and a pause:
```C
const float tone_length = 0.2f;
const float period = tone_length  * 8; // tone-gap-tone-gap-tone-gap-gap-gap = 8 beats
time = fmodf(time, period);
int beat = floorf(time / tone_length);
bool should_be_playing = (beat % 2) == 0 && beat != 6;
```

I hope this makes sense to you, if not let me know and I'll try to come up with a PoC to demonstrate better.

File: src/haptic_feedback.c (hunk 4 above)
Comment (Relys) [reply src/haptic_feedback.c:112]: If you don't immediately start playing a tone when called, then by the time you get to the haptic_feedback_update function timer and alt_timer will already be out of date.

File: src/haptic_feedback.c (hunk 5 above)
Comment (Relys) [reply src/haptic_feedback.c:152]: So I don't call any of the functions redundantly, only every period when alt_timer expires. Regarding the example for duty tone pattern, I'm not sure that this logic would blend smoothly with how I'm implementing creating a solid tone for when we've reached the duty cycle threshold (by continue to update alt_timer so it never reaches its next period). If you can provide an alternative that's fine, but I'm not sure I want to spend more cycles here rewriting this.

**Comment by Jared-Is-Coding** [comment]
I would suggest these revisions. GitHub isn't letting me "review", so I'll just post the commit:

https://github.com/Jared-Is-Coding/refloat/commit/be7611a893b6ca8d1ce312d3bc3f813f1e88fb64

Will give this some actual ride testing this week.

**Pull request closed without merging**